### ISSUE TITLE
[5.3] Removal of pcntl_fork() in queue-worker broke jobs with static properties

### ISSUE THREAD

**Issue by jaylinski** [laravel/framework#16694]
- Laravel Version: 5.3.26
- PHP Version: 7.1.0

### Description:

`pcntl_fork()` was removed with 23c12765557ebc5e3c35ad024d645620f7b907d6.

If jobs with static properties are executed in a forked process, the classes are always instantiated with a clean state.

By removing the process-forking, static vars are not reset when executing multiple jobs in a worker-process.

#### Test files
A class with static properties:
```php
<?php
namespace App;

class TestClass
{
    public static $testVar = 'original value';

    public function setTestVar($value)
    {
        self::$testVar = $value;
    }

    public function getTestVar()
    {
        return self::$testVar;
    }
}
```
A job that executes that class:
```php
<?php
namespace App\Jobs;

use App\TestClass;
use Illuminate\Bus\Queueable;
use Illuminate\Queue\SerializesModels;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Contracts\Queue\ShouldQueue;

class Test implements ShouldQueue
{
    use InteractsWithQueue, Queueable, SerializesModels;

    public function __construct()
    {
    }

    /**
     * Execute the job.
     */
    public function handle()
    {
        $testClass = new TestClass();
        echo $testClass->getTestVar().PHP_EOL;

        $testClass->setTestVar('changed value');

        echo $testClass->getTestVar().PHP_EOL;
    }
}
```
The command to dispatch the jobs:
```php
use App\Jobs\Test;
Artisan::command('pcntl_fork', function () {
    dispatch(new Test());
});
```
Call command multiple times to dispatch jobs:
```bash
php artisan pcntl_fork
php artisan pcntl_fork
php artisan pcntl_fork
```
Run the worker:
```bash
php artisan queue:work --queue=default
```

#### Output with Laravel Version 5.3.9
```
original value
changed value
[2016-12-07 13:43:41] Processed: App\Jobs\Test
original value
changed value
[2016-12-07 13:43:41] Processed: App\Jobs\Test
original value
changed value
[2016-12-07 13:43:42] Processed: App\Jobs\Test
```

#### Output with Laravel Version 5.3.26
```
original value
changed value
[2016-12-07 13:41:14] Processed: App\Jobs\Test
changed value
changed value
[2016-12-07 13:41:14] Processed: App\Jobs\Test
changed value
changed value
[2016-12-07 13:41:14] Processed: App\Jobs\Test
```

Kind of a breaking change.

I know that forking was removed because of SQS-issues.

Do you know of an easy workaround?

If there is no easy workaround, an additional `queue:work` flag could be added.
Something like `--fork-process`.

**Comment by dwightwatson** [comment]
Isn't this a bugfix if anything? I would have expected that static properties of a class would be reset between jobs, much like they would between web requests.

**Comment by jaylinski** [comment]
@dwightwatson Yep, I would also expect jobs to be executed like web requests.

There are a few possible solutions:
* Keep as is (and inform users about the new behavior in the docs).
* Add a `--fork-process` option to workers.
* Add a `private $fork = true|false;` variable to jobs, so it is possible to turn on/off process forking on job-level.
* Reintroduce process-forking and drop support for use-cases were SQS will fail.

I could try and submit a PR, but first I would like to know which solution the Laravel-Team would prefer.

**Comment by jaylinski** [comment]
Ping @taylorotwell Any input on this?

**Comment by wouthoekstra** [comment]
Also curious about developments on this. I've encountered unexpected changes of behavior since this change, breaking the handling of dispatched commands in queued jobs. A problem that remains in Laravel 5.4: https://github.com/laravel/framework/issues/18640

**Comment by jaylinski** [comment]
Won't be fixed as stated by @taylorotwell in PR #18711.

Possible alternative is to use something like https://github.com/mjphaynes/php-resque.

**Comment by humbcezar** [comment]
Another alternative is to use a singleton instead of static properties.
